src/server/server_family.cc

```diff
@@ -1629,11 +1631,65 @@ void ServerFamily::Memory(CmdArgList args, ConnectionContext* cntx) {
   return mem_cmd.Run(args);
 }
 
+void ServerFamily::BgSaveFb(bool new_version, std::string basename,
+                            boost::intrusive_ptr<Transaction> trans) {
+  GenericError ec = DoSave(new_version, basename, trans.get());
```

Comment:
I think that DoSave should be splited to the checks and init phase and to the wait and finish phase.
This way when you call bgsave you will return error imidiatly if there is some params error or save is already in progress (not only bgsave in progress).
And the bg save fiber will contain only the wait and finish phase